src/sqlancer/common/oracle/NoRECOracle.java

```diff
@@ -0,0 +1,142 @@
+package sqlancer.common.oracle;
+
+import java.sql.SQLException;
+import java.util.Arrays;
+
+import sqlancer.IgnoreMeException;
+import sqlancer.Randomly;
+import sqlancer.SQLGlobalState;
+import sqlancer.common.ast.newast.Expression;
+import sqlancer.common.ast.newast.Join;
+import sqlancer.common.ast.newast.Select;
+import sqlancer.common.gen.NoRECGenerator;
+import sqlancer.common.query.ExpectedErrors;
+import sqlancer.common.query.SQLQueryAdapter;
+import sqlancer.common.query.SQLancerResultSet;
+import sqlancer.common.schema.AbstractSchema;
+import sqlancer.common.schema.AbstractTable;
+import sqlancer.common.schema.AbstractTableColumn;
+import sqlancer.common.schema.AbstractTables;
+
+public class NoRECOracle<J extends Join<E, T, C>, E extends Expression<C>, S extends AbstractSchema<?, T>, T extends AbstractTable<C, ?, ?>, C extends AbstractTableColumn<?, ?>, G extends SQLGlobalState<?, S>>
+        implements TestOracle<G> {
+
+    private final G state;
+
+    private NoRECGenerator<J, E, T, C> gen;
+    private final ExpectedErrors errors;
+
+    private String lastQueryString;
+
+    public NoRECOracle(G state, NoRECGenerator<J, E, T, C> gen, ExpectedErrors expectedErrors) {
+        this.state = state;
```

Comment:
Given that this class will be widely used, perhaps we can implement it quite defensively? For example, here, we could throw an `IllegalArgumentException` if any of the arguments is `null` (I checked for other opportunities below, but didn't find any).

Reply:
good idea! I've made the changes